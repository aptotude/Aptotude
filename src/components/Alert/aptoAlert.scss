@import '../../scss/theme/variables.scss';

.AptoAlert {
  font-family: $font-family-base;
  display: block;
  line-height: $line-height-base;
  margin: 0 0 2rem 0;
  color: $body-text;
  padding: 1rem;
  border: 1px solid $primary-blue;
  background-color: rgba($primary-blue, 0.1);
  &.AptoAlert--warning {
    border-color: $warning;
    background-color: rgba($warning, 0.1);
  }
  &.AptoAlert--success {
    border-color: $success;
    background-color: rgba($success, 0.1);
  }
  &.AptoAlert--danger {
    border-color: $danger;
    background-color: rgba($danger, 0.1);
  }
}
