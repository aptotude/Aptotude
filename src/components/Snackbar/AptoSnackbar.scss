.AptoSnackbar {
  left: 0;
  right: 0;
  z-index: 1400;
  display: flex;
  position: fixed;
  align-items: center;
  justify-content: center;
  padding: 1rem;

  &.AptoSnackbarAnchorOriginbottomleft {
    justify-content: flex-start;
    bottom: 0;
    right: auto;
  }

  &.AptoSnackbarAnchorOriginbottomright {
    justify-content: flex-end;
    bottom: 0;
    left: auto;
  }

  &.AptoSnackbarAnchorOriginbottomcenter {
    justify-content: center;
    bottom: 0;
  }

  &.AptoSnackbarAnchorOrigintopleft {
    justify-content: flex-start;
    top: 0;
    right: auto;
  }

  &.AptoSnackbarAnchorOrigintopright {
    justify-content: flex-end;
    top: 0;
    left: auto;
  }

  &.AptoSnackbarAnchorOrigintopcenter {
    justify-content: center;
    top: 0;
  }
}
