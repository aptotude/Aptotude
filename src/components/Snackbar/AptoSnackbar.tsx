import * as React from 'react';
import { AptoSnackbarItem } from '../SnackbarItem/AptoSnackbarItem';
import { AptoSnackbarContext } from './AptoSnackbarContext';
import classNames from 'classnames';
import { StandardTypes } from '../../utils/standardTypes';
import { AptoButton } from '../Button';
import { AptoIcon } from '../Icon';
import { close } from '@apto/icons';

export type AptoSnackbarHorizontal = 'left' | 'center' | 'right';
export type AptoSnackbarVertical = 'top' | 'bottom';

export interface AptoSnackbarProps extends StandardTypes {
  anchorOrigin?: {
    horizontal: string | AptoSnackbarHorizontal;
    vertical: string | AptoSnackbarVertical;
  };
  duration?: number;
}

export class AptoSnackbar extends React.Component<AptoSnackbarProps> {
  public static defaultProps = {
    duration: 5000
  };

  public render() {
    const { className, duration } = this.props;
    const vertical =
      (this.props.anchorOrigin && this.props.anchorOrigin.vertical) || 'bottom';
    const horizontal =
      (this.props.anchorOrigin && this.props.anchorOrigin.horizontal) || 'left';
    const anchorClasses = classNames(
      `AptoSnackbar`,
      `AptoSnackbarAnchorOrigin${vertical}${horizontal}`,
      className
    );
    return (
      <AptoSnackbarContext.Consumer>
        {({ snacks, hideSnack }) => (
          <React.Fragment>
            {(snacks.length && (
              <div className={anchorClasses}>
                <div>
                  {snacks.map(snack => (
                    <AptoSnackbarItem
                      key={snack.id}
                      open={true}
                      variant={snack.type}
                      onClose={() => hideSnack(snack, null)}
                      autoHideDuration={duration}
                      action={
                        snack.action || [
                          <AptoButton
                            title="close"
                            kind="link"
                            variant="secondaryDark"
                            key="close"
                            data-testid="close-snack"
                            onClick={e => hideSnack(snack, e)}
                          >
                            <AptoIcon inline={true} icon={close} />
                          </AptoButton>
                        ]
                      }
                      message={snack.message}
                    />
                  ))}
                </div>
              </div>
            )) ||
              null}
          </React.Fragment>
        )}
      </AptoSnackbarContext.Consumer>
    );
  }
}
