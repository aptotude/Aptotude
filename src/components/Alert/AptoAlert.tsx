import React from 'react';
import classNames from 'classnames';
import { StandardTypes } from '../../utils/standardTypes';
import './aptoAlert.scss';

const COMPONENT_PREFIX = 'AptoAlert';

export interface AptoAlertDisplayProps extends StandardTypes {
  variant?: 'info' | 'success' | 'warning' | 'danger';
}

export class AptoAlert extends React.Component<AptoAlertDisplayProps> {
  public static defaultProps = {
    variant: 'info'
  };

  public render() {
    const { variant, className, forwardRef, children, ...rest } = this.props;

    const classes = classNames(
      COMPONENT_PREFIX,
      variant && `${COMPONENT_PREFIX}--${variant}`,
      className
    );

    return (
      <div {...rest} ref={forwardRef} className={classes}>
        {children}
      </div>
    );
  }
}
