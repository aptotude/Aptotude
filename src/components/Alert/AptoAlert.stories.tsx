import React from 'react';
import { storiesOf } from '@storybook/react';
import { AptoAlert } from './AptoAlert';

storiesOf('Alerts', module).add('Info', () => (
  <div>
    <AptoAlert>Info</AptoAlert>
    <AptoAlert variant="warning">Warning</AptoAlert>
    <AptoAlert variant="danger">Danger</AptoAlert>
    <AptoAlert variant="success">Success</AptoAlert>
  </div>
));
