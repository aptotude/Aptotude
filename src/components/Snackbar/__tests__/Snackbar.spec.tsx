import React from 'react';
import {
  render,
  cleanup,
  getByText,
  fireEvent,
  wait,
  getByTestId
} from 'react-testing-library';
import { AptoSnackbar, AptoSnackbarProps } from '../AptoSnackbar';
import {
  AptoSnackbarContext,
  AptoSnack,
  AptoSnackbarContextInterface
} from '../AptoSnackbarContext';
import { AptoAlertVariants } from '../../Alert';

let snackCounter = 0;

class TestComponent extends React.Component<
  AptoSnackbarProps,
  AptoSnackbarContextInterface
> {
  public showSnack = (
    message: React.ReactNode,
    type: AptoAlertVariants = 'info',
    action: React.ReactNode | null = null
  ) => {
    const snack: AptoSnack = {
      message,
      action,
      type,
      id: snackCounter += 1
    };
    const oldSnacks: AptoSnack[] = this.state.snacks || [];
    this.setState({ snacks: oldSnacks.concat(snack) });
  };

  public hideSnack = (removedSnack: AptoSnack) => {
    const oldSnacks: AptoSnack[] = this.state.snacks || [];
    const newSnacks = oldSnacks.filter(snack => snack.id !== removedSnack.id);
    this.setState({ snacks: newSnacks });
  };

  public state = {
    snacks: [],
    hideSnack: this.hideSnack,
    showSnack: this.showSnack
  };

  public render() {
    const { anchorOrigin } = this.props;
    let snackProps: AptoSnackbarProps = {};

    if (anchorOrigin) {
      snackProps = {
        anchorOrigin
      };
    }

    return (
      <AptoSnackbarContext.Provider value={this.state}>
        <AptoSnackbar {...snackProps} />
        <AptoSnackbarContext.Consumer>
          {snackContext => (
            <div>
              <button
                onClick={() => {
                  snackContext.showSnack(`This is a snack`);
                }}
              >
                Open Snack
              </button>
              <button
                onClick={() => {
                  snackContext.showSnack(`This is another snack`, 'danger');
                }}
              >
                Open Another Snack
              </button>
            </div>
          )}
        </AptoSnackbarContext.Consumer>
      </AptoSnackbarContext.Provider>
    );
  }
}

describe('Snackbar Component', () => {
  afterEach(cleanup);

  it('It renders', () => {
    const { container } = render(<TestComponent />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('It adds a snack', async () => {
    const { container } = render(<TestComponent />);

    const btn = getByText(container, 'Open Snack');

    fireEvent.click(btn);

    await wait(() => {
      expect(container.querySelector('.AptoSnackbar')).toMatchSnapshot();
    });
  });

  it('It adds multiple snacks', async () => {
    const { container } = render(<TestComponent />);

    const btn = getByText(container, 'Open Snack');
    const btn2 = getByText(container, 'Open Another Snack');

    fireEvent.click(btn);
    fireEvent.click(btn2);

    await wait(() => {
      expect(container.querySelector('.AptoSnackbar')).toMatchSnapshot();
    });
  });

  it('It Closes a snack when clicking close', async () => {
    const { container } = render(<TestComponent />);

    const btn = getByText(container, 'Open Snack');

    fireEvent.click(btn);

    await wait(() => {
      const close = getByTestId(container, 'close-snack');
      fireEvent.click(close);
      expect(container.querySelector('.AptoSnackbar')).toMatchSnapshot();
    });
  });

  it('It lets you set anchor location', async () => {
    const props = {
      anchorOrigin: {
        vertical: 'top',
        horizontal: 'right'
      }
    };

    const { container } = render(<TestComponent {...props} />);

    const btn = getByText(container, 'Open Snack');

    fireEvent.click(btn);

    await wait(() => {
      expect(container.querySelector('.AptoSnackbar')).toMatchSnapshot();
    });
  });
});
