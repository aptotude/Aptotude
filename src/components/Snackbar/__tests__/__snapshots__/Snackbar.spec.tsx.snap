// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar Component It Closes a snack when clicking close 1`] = `null`;

exports[`Snackbar Component It adds a snack 1`] = `
<div
  class="AptoSnackbar AptoSnackbarAnchorOriginbottomleft"
>
  <div>
    <div
      class="AptoSnackbarItem"
    >
      <div
        class="AptoAlert AptoAlert--info AptoSnackbarItemContent"
      >
        <div
          class="AptoSnackbarItemContentMessage"
        >
          This is a snack
        </div>
        <div
          class="AptoSnackbarItemContentAction"
        >
          <button
            aria-label="close"
            class="AptoButton AptoButton--link AptoButton--secondaryDark"
            data-testid="close-snack"
          >
            <span
              class="AptoIcon AptoIcon--size3 AptoIcon--inline"
            >
              <svg
                focusable="false"
                height="100%"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 24 24"
                width="100%"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snackbar Component It adds multiple snacks 1`] = `
<div
  class="AptoSnackbar AptoSnackbarAnchorOriginbottomleft"
>
  <div>
    <div
      class="AptoSnackbarItem"
    >
      <div
        class="AptoAlert AptoAlert--info AptoSnackbarItemContent"
      >
        <div
          class="AptoSnackbarItemContentMessage"
        >
          This is a snack
        </div>
        <div
          class="AptoSnackbarItemContentAction"
        >
          <button
            aria-label="close"
            class="AptoButton AptoButton--link AptoButton--secondaryDark"
            data-testid="close-snack"
          >
            <span
              class="AptoIcon AptoIcon--size3 AptoIcon--inline"
            >
              <svg
                focusable="false"
                height="100%"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 24 24"
                width="100%"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="AptoSnackbarItem"
    >
      <div
        class="AptoAlert AptoAlert--danger AptoSnackbarItemContent"
      >
        <div
          class="AptoSnackbarItemContentMessage"
        >
          This is another snack
        </div>
        <div
          class="AptoSnackbarItemContentAction"
        >
          <button
            aria-label="close"
            class="AptoButton AptoButton--link AptoButton--secondaryDark"
            data-testid="close-snack"
          >
            <span
              class="AptoIcon AptoIcon--size3 AptoIcon--inline"
            >
              <svg
                focusable="false"
                height="100%"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 24 24"
                width="100%"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snackbar Component It lets you set anchor location 1`] = `
<div
  class="AptoSnackbar AptoSnackbarAnchorOrigintopright"
>
  <div>
    <div
      class="AptoSnackbarItem"
    >
      <div
        class="AptoAlert AptoAlert--info AptoSnackbarItemContent"
      >
        <div
          class="AptoSnackbarItemContentMessage"
        >
          This is a snack
        </div>
        <div
          class="AptoSnackbarItemContentAction"
        >
          <button
            aria-label="close"
            class="AptoButton AptoButton--link AptoButton--secondaryDark"
            data-testid="close-snack"
          >
            <span
              class="AptoIcon AptoIcon--size3 AptoIcon--inline"
            >
              <svg
                focusable="false"
                height="100%"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 24 24"
                width="100%"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snackbar Component It renders 1`] = `
<div>
  <button>
    Open Snack
  </button>
  <button>
    Open Another Snack
  </button>
</div>
`;
